op-node/rollup/finality/finalizer.go

```diff
@@ -236,33 +245,52 @@ func (fi *Finalizer) tryFinalize() {
 	// go through the latest inclusion data, and find the last L2 block that was derived from a finalized L1 block
 	for _, fd := range fi.finalityData {
 		if fd.L2Block.Number > finalizedL2.Number && fd.L1Block.Number <= fi.finalizedL1.Number {
-			// check if fd.L2Block.Number is finalized on Babylon
-			queryParams := &sdk.L2Block{
-				BlockHeight:    fd.L2Block.Number,
-				BlockHash:      fd.L2Block.Hash.String(),
-				BlockTimestamp: fd.L2Block.Time,
+			// If there are any L2 blocks that are not finalized, then we can't finalize the later ones.
+			// This is because we need to finalize the L2 blocks in order.
+			blockCount := int(fd.L2Block.Number - finalizedL2.Number)
+			l2Blocks := make(map[uint64]eth.L2BlockRef)
+			queryBlocks := make([]*cwclient.L2Block, blockCount)
+			for i := 0; i < blockCount; i++ {
+				blockNumber := uint64(i) + finalizedL2.Number + uint64(1)
+				l2Block, err := fi.l2Fetcher.L2BlockRefByNumber(fi.ctx, blockNumber)
+				if err != nil {
+					fi.emitter.Emit(rollup.EngineTemporaryErrorEvent{Err: fmt.Errorf("failed to check if on finalizing L2 chain, could not fetch block %d: %w", blockNumber, err)})
+					return
+				}
+				l2Blocks[blockNumber] = l2Block
+
+				queryBlocks[i] = &cwclient.L2Block{
+					BlockHeight:    l2Block.Number,
+					BlockHash:      l2Block.Hash.String(),
+					BlockTimestamp: l2Block.Time,
+				}
+				fi.log.Debug(
+					"babylon gadget query params",
+					"block_height", queryBlocks[i].BlockHeight,
+					"block_hash", queryBlocks[i].BlockHash,
+					"block_timestamp", queryBlocks[i].BlockTimestamp,
+				)
+
 			}
-			fi.log.Debug(
-				"babylon gadget query params",
-				"block_height", queryParams.BlockHeight,
-				"block_hash", queryParams.BlockHash,
-				"block_timestamp", queryParams.BlockTimestamp,
-			)
-			babylonFinalized, err := fi.babylonFinalityClient.QueryIsBlockBabylonFinalized(queryParams)
-			fi.log.Debug("babylon gadget query result", "babylon_finalized", babylonFinalized)
+			// query the last finalized block in the block range on Babylon
+			lastFinalizedBlockNumber, err := fi.babylonFinalityClient.QueryBlockRangeBabylonFinalized(queryBlocks)
+			fi.log.Debug("babylon gadget query result", "babylon_finalized_block_number", lastFinalizedBlockNumber)
 
 			// If the error encountered is of type NoFpHasVotingPowerError, it should be ignored;
 			// for any other error types, emit a critical error event.
-			if err != nil && !errors.Is(err, sdk.ErrNoFpHasVotingPower) {
+			if err != nil && !errors.Is(err, sdkclient.ErrNoFpHasVotingPower) {
```

Comment:
thinking deeper on it, this check is wrong. see https://github.com/babylonchain/babylon-finality-gadget/issues/59

this is non-trivial change. so likely will need another PR